pallets/block-rewards/src/lib.rs

```diff
@@ -0,0 +1,475 @@
+			origin: OriginFor<T>,
+			collator_reward_per_session: T::Balance,
+		) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			NextSessionChanges::<T>::mutate(|changes| {
+				changes.collator_reward = Some(collator_reward_per_session);
+			});
+
+			Ok(())
+		}
+
+		/// Admin method to set the total reward distribution for the next sessions.
+		/// Current session is not affected by this call.
+		///
+		/// Throws if total_reward < collator_reward * collator_count.
+		#[pallet::weight(T::WeightInfo::set_total_reward())]
+		pub fn set_total_reward(
+			origin: OriginFor<T>,
+			total_reward_per_session: T::Balance,
+		) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			NextSessionChanges::<T>::try_mutate(|changes| {
+				let current = ActiveSessionData::<T>::get();
+				let total_collator_rewards = changes
+					.collator_reward
+					.unwrap_or(current.collator_reward)
+					.saturating_mul(
+						changes
+							.collator_count
+							.unwrap_or(current.collator_count)
+							.into(),
+					);
+				ensure!(
+					total_reward_per_session >= total_collator_rewards,
+					Error::<T>::InsufficientTotalReward
+				);
+
+				changes.total_reward = Some(total_reward_per_session);
+				Ok(())
+			})
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	/// Mint default amount of stake for target address and deposit stake.
+	/// Enables receiving rewards onwards.
+	///
+	/// Weight (6 reads, 6 writes):
+	///  * mint_into (2 reads, 2 writes): Account, TotalIssuance
+	///  * deposit_stake (4 reads, 4 writes): Currencies, Groups, StakeAccounts, Account
+	pub(crate) fn do_init_collator(who: &T::AccountId) -> DispatchResult {
+		T::Currency::mint_into(STAKE_CURRENCY_ID, who, DEFAULT_COLLATOR_STAKE.into())?;
+		T::Rewards::deposit_stake(
+			(T::Domain::get(), STAKE_CURRENCY_ID),
+			who,
+			DEFAULT_COLLATOR_STAKE.into(),
+		)
+	}
+
+	/// Withdraw currently staked amount for target address and immediately burn it.
+	/// Disables receiving rewards onwards.
+	pub(crate) fn do_exit_collator(who: &T::AccountId) -> DispatchResult {
+		let amount = T::Rewards::account_stake((T::Domain::get(), STAKE_CURRENCY_ID), who);
+		T::Rewards::withdraw_stake((T::Domain::get(), STAKE_CURRENCY_ID), who, amount)?;
+		T::Currency::burn_from(STAKE_CURRENCY_ID, who, amount).map(|_| ())
+	}
+
+	/// Apply session changes and distribute rewards.
+	///
+	/// NOTE: Noop if any call fails.
+	fn do_advance_session() {
+		let mut num_joining = 0u32;
+		let mut num_leaving = 0u32;
+
+		transactional::with_storage_layer(|| -> DispatchResult {
+			NextSessionChanges::<T>::try_mutate(|changes| -> DispatchResult {
+				ActiveSessionData::<T>::try_mutate(|session_data| {
+					// Reward collator group of last session
+					let total_collator_reward = session_data
+						.collator_reward
+						.saturating_mul(session_data.collator_count.into())
+						.min(session_data.total_reward);
+					T::Rewards::reward_group(COLLATOR_GROUP_ID, total_collator_reward)?;
```

Comment:
I'm wondering if `COLLATOR_GROUP_ID` could be used by another pallet. Maybe we need to use some "Domain" mechanism here too.

Reply:
The admin of the liquidity pallet can create a group; if they choose the same ID as this, what would happen?

Reply:
AFAIU, `pallet-liquidity-rewards` and `pallet-block-rewards` each require a separate instance of `pallet-rewards`. Thus, even if the same GroupID exists in `pallet-liquidity-rewards`, it cannot influence `pallet-block-rewards`.

Reply:
I definitely agree this could potentially lead to a silent bug. However, what do we win if we define it in `Config`?

Reply:
Visibility. it's no longer an internal library value. However, it's true that we should configure some intervals to use in the liquidity, too to avoid collisions. So we do not win anything 😅. We need some kind of group domain level or always use a different `pallet-reward` each time.

Maybe we can remove the `Domain` concept that adds a lot of noise and always forces us to use different instances. This would simplify things a lot 🤔. What do you think about this? Of course, this will be another PR.

Reply:
The `Domain` concept was something that we implemented before allowing `pallet-rewards` to have several instances. So maybe this is now a concept that can be deprecated

Reply:
> The Domain concept was something that we implemented before allowing pallet-rewards to have several instances. So maybe this is now a concept that can be deprecated

Thanks for the insights. I was wondering whether it could be simplified but did not want to add noise at the current stage.

> Maybe we can remove the `Domain` concept that adds a lot of noise and always forces us to use different instances. This would simplify things a lot 🤔. What do you think about this? Of course, this will be another PR.

I agree, will open a ticket!

Reply:
https://github.com/centrifuge/centrifuge-chain/issues/1246

Reply:
> I was wondering whether it could be simplified, but I did not want to add noise at the current stage.

Do it, do it. It is needed to avoid currency collisions using the same `pallet-rewards`. But I can be blind to one solution because I'm too biased on this. New mind sets are welcome!